### PR TITLE
Distributed data processing

### PR DESCRIPTION
**Pull request by jychoi-hpc**
This is the update I made to run Ising model using multi-GPUs on Summit and Perlmutter.

Major updates:
* Add distributed raw data/serial data loading (each process will read different set of data)
* Parallel normalization and min-max (using DDP to aggregate)
* Using Adios2 for Ising

It looks like many changes but I think this is necessary for distributed processing. Please feel free to give any comments.

### CODE REVIEW

**Review by streeve — COMMENTED** [review]
Like we discussed at the meeting we can switch to DDP required an simplify some of these changes (could be separate if you prefer)

File: hydragnn/models/Base.py [hunk 1]
```diff
@@ -226,6 +228,7 @@ def _multihead(self):
             self.heads_NN.append(head_NN)
 
     def forward(self, data):
+        data = data.to(self.device)
```
[the remark this answers is not in the document]
Comment (jychoi-hpc) [reply hydragnn/models/Base.py:239]: I found previously we move data from cpu to gpu in serialized data loader: https://github.com/ORNL/HydraGNN/blob/b48f75cd3fe1b0d03bae9af3e6bdc2bb29f8b9c6/hydragnn/preprocess/serialized_dataset_loader.py#L140
I think moving data from cpu to gpu here makes more sense to me. It also works with the case when serialized loader is not used (like Adios).

File: examples/ising_model/create_configurations.py [hunk 2]
```diff
@@ -43,9 +43,9 @@ def E_dimensionless(config, L, spin_function, scale_spin):
     number_nodes = L ** 3
     positions = np.zeros((number_nodes, 3))
     atomic_features = np.zeros((number_nodes, 5))
-    for z in range(L):
+    for x in range(L):
```
Comment (allaffa) [inline examples/ising_model/create_configurations.py:46]: Can you explain why you had to invert the order of these nested loop?

File: hydragnn/models/Base.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (allaffa) [reply hydragnn/models/Base.py:239]: Previously, @streeve put this line inside the serialized data loader because small data can fit entirely on the GPU. 
I am on to lose some accuracy on small data by benefitting on large data that needs HPC resources. Because there is where we really care about performance. 
@pzhanggit what do you think?

File: hydragnn/preprocess/load_data.py [hunk 3]
```diff
@@ -50,15 +52,28 @@ def dataset_loading_and_splitting(config: {}):
     )
 
 
+def worker_init_fn(worker_id):
```
Comment (allaffa) [inline hydragnn/preprocess/load_data.py:55]: Did you add this function with the idea that the functionalities of this function could be expanded in the future? 
It seems like the only thing it does is writing a line on the log file. Is it true?

File: hydragnn/preprocess/utils.py [hunk 4]
```diff
@@ -16,16 +16,54 @@
 import ase
 import ase.neighborlist
 
-
+## This function can be slow if dataset is too large. Use with caution.
```
Comment (allaffa) [inline hydragnn/preprocess/utils.py:19]: Is this where @streeve was mentioning the idea of avoiding multiple replicas of the same function?

File: hydragnn/utils/model.py [hunk 5]
```diff
@@ -90,14 +81,42 @@ def load_existing_model(model, model_name, path="./logs/", optimizer=None):
         optimizer.load_state_dict(checkpoint["optimizer_state_dict"])
 
 
+## This function may cause OOM if dataset is too large
+## to fit in a single GPU (i.e., with DDP). Use with caution.
+## Recommend to use calculate_PNA_degree_dist
 def calculate_PNA_degree(dataset: [Data], max_neighbours):
-    deg = torch.zeros(max_neighbours + 1, dtype=torch.long).to(get_device())
+    deg = torch.zeros(max_neighbours + 1, dtype=torch.long)
```
Comment (allaffa) [inline hydragnn/utils/model.py:88]: I assume this is another point where @streeve was mentioning the idea of avoiding function replicas. Correct?

**Comment by jychoi-hpc** [comment]
Let's drop this PR for now. Will make a PR with the current OGB/CSCE update.
